### PR TITLE
Add X11 Primary Selection clipboard support

### PR DESCRIPTION
**Pull request by davidemdot**
_Related to issue #2389 (**JabRef and the X Window clipboard/selections**)._

I have been testing different alternatives and I think that this is the best choice (for using `XClipboard` a new dependency is needed, and it does not give any added value).

- All contents that JabRef sends to the system clipboard will be available in the X11 Primary clipboard too (i.e., `copyVersionToClipboard()` in the About dialog). So it is possible to paste it in a text editor, doing click in the middle/third mouse button.

copy_version

- It is necessary to call the `install()` method in every text field where you want to use it. For the moment, it is activated in the `GlobalSearchBar`, therefore you can try it: write something, select a piece of text and do click in the middle button.

global_search

- Now `ClipBoardManager` uses the singleton pattern. It is instantiated once in `Globals.java`.

Any feedback will be more than welcome! **:"D**

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

**Comment by koppor** [comment]
Thank you on working on this. No Linux at hand currently.

The primary and secondary buffers are completly separate buffers. So, the test case is:

1. Open JabRef
2. Add a new entry
3. Have the entry editor opened, showing the author and title field
4. Select `Milton {Abramowitz} and Irene A. {Stegun}` (here in this issue) and press <kbd>Ctrl</kbd>+<kbd>C</kbd> to copy that text to the primary buffer
5. Select `Handbook of Mathematical Functions with Formulas, Graphs, and Mathematical Tables` with the mouse. To not press any keyboard shortcut
6. Switch to JabRef (maybe using <kdbd>Alt</kbd>+<kbd>Tab</kbd>)
7. Go to the title field.
8. Press the middle mouse key.
9. Check expectation: `Handbook of Mathematical Functions with Formulas, Graphs, and Mathematical Tables` is pasted.
10. Go to the author field.
11. Press <kbd>Ctrl</kbd>+<kbd>V</kbd>
12. Expectation: `Milton {Abramowitz} and Irene A. {Stegun}` appears.

@gregoa This is the expected behavior if I remember the Linux desktop correctly. I checked https://docs.kde.org/trunk5/en/kde-workspace/klipper/clipboard-modes.html, which seems a good short resource on that. I see at https://ask.slashdot.org/story/04/06/02/1832201/dealing-with-the-unix-copy-and-paste-paradigm that Firefox pastes the secondary clipboard at <kbd>Ctrl</kbd>+<kbd>V</kbd> if text was selected in firefox.

Short summary for X11 clipbaords: https://unix.stackexchange.com/a/139193/18033. Longer text (good introd pargraphs) : https://www.uninformativ.de/blog/postings/2017-04-02/0/POSTING-en.html. I learned, there is `CLIPBOARD` (typically <kbd>Ctrl</kbd>+<kbd>C</kbd>/<kbd>V</kbd>) and `PRIMARY` (mouse text highlight).

Source for the bibtex example: https://en.wikipedia.org/wiki/BibTeX

**Comment by davidemdot** [comment]
@koppor After last commit, which activates the `PRIMARY` clipboard for all `TextField`/`TextArea` controls in the entry editor, your test case works exactly as you said.

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 1]
```diff
@@ -30,34 +37,56 @@
 import org.slf4j.LoggerFactory;
 
 public class ClipBoardManager {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: A better approach for singleton is to use an enum https://stackoverflow.com/a/26285591/3450689

Comment (koppor) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: Source of the information: [Effective Java Item 3](https://www.oreilly.com/library/view/effective-java-3rd/9780134686097/). A highly-recommended reading. A summary of the book can be found at https://github.com/HugoMatilla/Effective-JAVA-Summary. (Just commented at stackoverflow - thank you for the pointer)

**Comment by koppor** [comment]
@gregoa @noMICROSOFTbuhtz Could you test if it meets your expectations?

File: src/main/java/org/jabref/gui/ClipBoardManager.java (hunk 1 above)
Comment (davidemdot) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: Thanks, @Siedlerchr!

File: src/main/java/org/jabref/Globals.java [hunk 2]
```diff
@@ -56,13 +56,11 @@
...
      * Manager for the state of the GUI.
      */
-
-    public static ClipBoardManager clipboardManager = new ClipBoardManager();
-
     public static StateManager stateManager = new StateManager();
     public static ExporterFactory exportFactory;
     public static CountingUndoManager undoManager = new CountingUndoManager();
     public static BibEntryTypesManager entryTypesManager = new BibEntryTypesManager();
+    public static ClipBoardManager clipboardManager = ClipBoardManager.INSTANCE;
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/Globals.java]: That's a bit of a controversial topic (also among JabRef's devs) but we try to avoid to use the `Singleton` pattern now. It makes it harder to test code as you cannot mock or replace the instance (as it's a final field and the constructor private). Moreover, the code should always use the manager defined here in `Globals` instead of `Clipboardmanager.instance` and there is no Java language feature that helps us here. 

So in summary I would like to ask you to revert this change.

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 3]
```diff
@@ -29,35 +34,43 @@
...
-    /**
-     * Puts content onto the clipboard.
-     */
-    public void setContent(ClipboardContent content) {
-        clipboard.setContent(content);
+    public void inputToPrimary(TextInputControl input) {
+        input.insertText(input.getCaretPosition(), getContentsPrimary());
     }
 
+    // For activating this functionality in an input, use:
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: It's nice that you provide code examples and the Javadoc has the `{@code` to make this easier (and in a standardized way)

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 4]
```diff
@@ -29,35 +34,43 @@
...
+    //
+    // input.selectedTextProperty().addListener((observable, oldValue, newValue) -> watchInput(newValue));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: It would be nice if these setup steps could be simplified and combined into one method, so that it is enough to call `ClipBoardManager.addX11Support(input)`. This also has the advantage that you can inline `watchInput` and `inputToPrimary`.

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 5]
```diff
@@ -67,16 +80,44 @@ public String getContents() {
         return result;
     }
 
+    // Get the text from Primary
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: It's java convention to use `/** ... */` for these comments. Search for "JavaDoc" if you want to learn more (and see also which tags are supported).

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 6]
```diff
@@ -67,16 +80,44 @@ public String getContents() {
...
+                return (String) contents.getTransferData(DataFlavor.stringFlavor);
+            } catch (UnsupportedFlavorException | IOException e) {
+                LOGGER.warn(e.getMessage());
+            }
+        }
+        return getContents();
+    }
+
+    // Copy from Clipboard to Primary
+    private void clipboardToPrimary(ClipboardContent content) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: What about renaming to `setPrimaryClipboardContent` ?

File: src/main/java/org/jabref/gui/ClipBoardManager.java (hunk 4 above)
Comment (davidemdot) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: Yep. I did something like that in the first version...

https://github.com/JabRef/jabref/blob/a54ddd174ac8c0f3906625cc77824fe1f382f68e/src/main/java/org/jabref/gui/ClipBoardManager.java#L73-L84

But then I spent many hours trying to solve some errors that appeared when doing `./gradlew test`. As I wanted to have an available _bugs free_ version, I made those changes; but I will keep improving and testing this stuff in the next few days.

**:"D**

File: src/main/java/org/jabref/Globals.java (hunk 2 above)
Comment (davidemdot) [reply src/main/java/org/jabref/Globals.java]: I will ll do, no prob. I supposed that it was a risky decision... :sweat_smile:

Comment (tobiasdiez) [reply src/main/java/org/jabref/Globals.java]: It's really good that you went for it! Learning by doing is the best approach for these software architecture questions.

File: src/main/java/org/jabref/gui/ClipBoardManager.java (hunk 4 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: That could also be a problem with our gui tests. There are a bit unreliably (to phrase it politely).

File: src/main/java/org/jabref/Globals.java (hunk 2 above)
Comment (davidemdot) [reply src/main/java/org/jabref/Globals.java]: Good good. I got a working snapshot without using the singleton pattern. I think the `addX11Support()` method should be static, so I also switched to static attributes (this way, we can access from static and non-static methods). I await your feedback!

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 7]
```diff
@@ -30,53 +37,108 @@
...
+    public ClipBoardManager(Clipboard clipboard, java.awt.datatransfer.Clipboard primary, ImportFormatReader importFormatReader) {
+        ClipBoardManager.clipboard = clipboard;
+        ClipBoardManager.primary = primary;
+        ClipBoardManager.importFormatReader = importFormatReader;
     }
 
-    public ClipBoardManager(Clipboard clipboard, ImportFormatReader importFormatReader) {
-        this.clipboard = clipboard;
-        this.importFormatReader = importFormatReader;
+    public ClipBoardManager() {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: Why were the two constructors swapped? Could you keep them in the same order as they were?

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 8]
```diff
@@ -30,53 +37,108 @@
...
+     */
+    public void setPrimaryClipboardContent(ClipboardContent content) {
+        primary.setContents(new StringSelection(content.getString()), null);
+    }
+
     public void setHtmlContent(String html) {
         final ClipboardContent content = new ClipboardContent();
         content.putHtml(html);
         clipboard.setContent(content);
+        setPrimaryClipboardContent(content);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: Is this line really necessary? I think, only when selecting text by the mouse should copy something. Not in all cases.

Maybe, I understand the code wrong. Currently, it reads that the content is copied into the system clipboard and the "primary" one (which is the one where the middle mouse pastes from)

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 9]
```diff
@@ -30,53 +37,108 @@
...
     }
 
     public void setContent(String string) {
         final ClipboardContent content = new ClipboardContent();
         content.putString(string);
         clipboard.setContent(content);
+        setPrimaryClipboardContent(content);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: See above. I don't think, we need this line.

File: src/main/java/org/jabref/gui/ClipBoardManager.java [hunk 10]
```diff
@@ -86,6 +148,7 @@ public void setContent(List<BibEntry> entries) throws IOException {
         content.put(DragAndDropDataFormats.ENTRIES, serializedEntries);
         content.putString(serializedEntries);
         clipboard.setContent(content);
+        setPrimaryClipboardContent(content);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/ClipBoardManager.java]: See above. I don't think, we need this line.

File: src/main/java/org/jabref/gui/ClipBoardManager.java (hunk 7 above)
Comment (davidemdot) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: D'oh! Sorry, I didn't realize that I made that change. I'll revert it.

File: src/main/java/org/jabref/gui/ClipBoardManager.java (hunk 8 above)
Comment (davidemdot) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: It is not necessary. I have included that call because in the original issue (#2389) the JabRef contributor said:

> in the "Copy DOI URL" function which copies the URL into the XA_CLIPBOARD selection (**and not (also) into XA_PRIMARY**)

So at the beginning I thought that we wanted to copy to both clipboards a specific content, for example, when a user do a click in the `Copy Version` button. But we could remove it, if you prefer that behavior.

Comment (koppor) [reply src/main/java/org/jabref/gui/ClipBoardManager.java]: OK, then leave it in.
